[READY] New nanite program in the Utility category, Nanite Sting.

Nebulacrity opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
It is a Trigger-based close-range nanite-spreading program. It requires the Military Nanite Programming research node. Targets of this program will notice that they got stung. If no valid targets are found, the used volume is refunded.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Nanites are quite underwhelming currently, and any new program makes them a little bit more interesting. Note that this program is a nanite-spreading one - the two only other ways of spreading nanites is to either shove people into Nanite Chambers or to get lucky and obtain an alien tool which allows the research of Infective Exo-Locomotion. Antagonists and non-antagonists alike would be a bit more drawn to nanites if they could spread their creations faster, with less manual input.

Additionally, the existence of this program would add some confusion to some scenarios - if someone notices that they were stung, they might start accusing people around them of being Changelings, or, if someone was stung by a Changeling, they might just think it's this new program.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Added Nanite Sting nanite program.
:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Nebulacrity committed:
Adds a new nanite program in the Utility category, Nanite Sting. It is a manual close-range nanite-spreading program.

Detective-Google commented:
There's a reason infective locomotion is locked behind alien tech, though this would be neat if it was also locked behind the same tech 👎

zxaber commented:
I agree, forcing nanites upon people is meh. Taking nanites at all is already a gamble that some shitter doesn't get into science and mess with the cloud ID, and this alone is probably why people don't bother with them. The last thing I want to deal with is some scientist being "helpful" and forcibly spreading a nanite swarm to the crew, and then Tator McTot waltzes in and just uploads the death programs.

If you _have_ to add this, at least make the sting a guaranteed message to the victim, and not just a 20% chance. People deserve to know when a fellow crewmember is openly putting them at risk of remote instadeath.

py01 commented:
Having a way to forcibly infect targets will definitely give a use to all the negative programs that are otherwise completely useless, however you may need to re-tune the suppression programs, to prevent this being used for stealthy long hard stuns.

One possible solution would be to give the sting a do_after so the target has to stand still next to someone, giving this a bit more room to avoid.

Nebulacrity committed:
Increased the sting's notification chance of appearing from 20% to 100%.

Nebulacrity reviewed (commented):
Due to feedback I have removed the random appearance chance of the sting notification which targets of Nanite Sting see. Now it should be way more noticeable, at least if the programmer makes it activate rapidly.

Nebulacrity commented:
> There's a reason infective locomotion is locked behind alien tech, though this would be neat if it was also locked behind the same tech 👎

Infective Exo-Locomotion is very powerful, stealthy and fast. Obviously it'd be locked in such a way. Nanite Sting is way worse in comparison, and it's pretty much impossible to have it spread nanites stealthily since it alerts the affected targets. I intentionally made this program so there'd be at least some non-IEL way of spreading nanites with little manual input.

Qustinnus commented on code/modules/research/nanites/nanite_programs/utility.dm:
```diff
@@ -258,6 +258,32 @@
+		if(!(L.mob_biotypes & (MOB_ORGANIC|MOB_UNDEAD)))
+			continue
+		target_hosts += L
+		if(!target_hosts.len)
```
This will never run.

Qustinnus replied:
you add something to the list before this one

Qustinnus replied:
nevermind you fucked up indentation, this is supposed to run outside of the for loop. Rifght now you run all of the code every time you loop through, which means you infect multiple people.

Nebulacrity committed:
Fixed incorrect syntax which caused an inactive bit of code.

Qustinnus commented on code/modules/research/nanites/nanite_programs/utility.dm:
```diff
@@ -258,6 +258,32 @@
+			continue
+		target_hosts += L
+	if(!target_hosts.len)
+		consume_nanites(-5)
```
i dont know much about nanites but shouldnt the nanites always be consumed?

Nebulacrity replied:
The nanite cluster in this case seeks for a target, but if it doesn't find any it returns to the host.

Qustinnus replied:
That makes sense.

Qustinnus commented on code/modules/research/nanites/nanite_programs/utility.dm:
```diff
@@ -258,6 +258,32 @@
+		infectee.AddComponent(/datum/component/nanites, 5)
+		SEND_SIGNAL(infectee, COMSIG_NANITE_SYNC, nanites)
+		infectee.investigate_log("was infected by a nanite cluster by [key_name(host_mob)] at [AREACOORD(infectee)].", INVESTIGATE_NANITES)
+		to_chat(infectee, "<span class='warning'>You feel a tiny prick.</span>")
```
there's not a chance they notice here, they always notice.

Nebulacrity replied:
So what will it be? Will people notice? I'd say a good 60% would. (that is, if you're referring to the tiny prick; otherwise, I'm not sure what you're talking about)

Qustinnus replied:
I'm saying that "Targets of this program have a chance to notice that they got stung." is not accurate; this code will always send a message to the infectee

Nebulacrity replied:
So you're saying I should correct that. Correct?

Nebulacrity reviewed (commented):
Fixed fucked up indentation. Thanks to Qustinnus for pointing it out.

Nebulacrity committed:
Edited the description of the program, clarifying that it will be felt by the target and that if there was no target found, the nanite volume cost is refunded.

ExcessiveUseOfCobblestone commented:
@XDTM thots?

my 2 cents is that this messes up the overarching balance of nanites which were suppose to be like surgery in the sense of "strong but required the patient to be willing", making it spread like a virus would mean each current program/combo would need to be reviewed.

Nebulacrity commented:
> my 2 cents is that this messes up the overarching balance of nanites which were suppose to be like surgery in the sense of "strong but required the patient to be willing", making it spread like a virus would mean each current program/combo would need to be reviewed.

Eh, I'd say this case is a grey area. It requires researching Smart Nanite Programming. If we look at the description of this research node, here's what's written: "Nanite programs that require nanites to perform complex actions, act independently, roam or seek targets." Which is exactly what's being performed by this program - a small group of nanites form a cluster (a semi-complex action), seek out a valid target near the host and enter said target's body. Fits quite well in my opinion.

Putnam3145 commented:
infective exo-locomotion is actually super weak, it's worse than sneezing and uses so many nanites that you can't keep it up all the time, this is going to be way stronger despite what you may think

Nebulacrity commented:
> infective exo-locomotion is actually super weak, it's worse than sneezing and uses so many nanites that you can't keep it up all the time, this is going to be way stronger despite what you may think

Obviously you're gonna run out of nanites if you keep IEL 24/7. There must be alternating burn & regeneration phases programmed to keep the volume at a good amount.

In comparison to Nanite Sting, IEL has a far larger range and is silent. These two things are the main reason why IEL was made so hard to research imo. 
If even one person notices the Nanite Sting's "little prick" notification, they'll likely call it out. If they'll call it out, people will be more attentive, thus further decreasing the already-miniscule chances of spreading the nanites stealthily. And the more people start noticing these notifications, the more likely they'll go raid Science and likely fuck with the Cloud Control Console(s).

XDTM commented:
Nanites are specifically balanced with the thought that the hosts are voluntary and took on the risks: some effects are equivalent to instadeath. For that reason i limited the only infection program to abductor-research-only, making it not always available and requiring late-game research.

I believe that thanks to Cobby's changes to medicine they're eventually going to look more attractive to people who want rapid healing, despite their risks, even without any direct changes.
If you want more nanite infection gameplay, i'd look into reducing exo-locomotion's cost or increasing the infection chance, since from the rounds i spectated it was quite underwhelming.

Putnam3145 commented:
> Obviously you're gonna run out of nanites if you keep IEL 24/7. There must be alternating burn & regeneration phases programmed to keep the volume at a good amount.

Sure, except my actual point was that it's not strong enough to infect *anyone* unless it's on constantly, which isn't possible as previously mentioned

Nebulacrity committed:
Fixed Nanite Sting taking other hosts into account.
Changed Nanite Sting's required Research Node from Smart Nanite Programming to Military Nanite Programming.
Changed cooldown from 3s to 10s.

Nebulacrity commented:
Due to feedback I have changed the required Research Node from Smart Nanite Programming to Military Nanite Programming increased the cooldown from 3s to 10s.
I have also fixed the ability targeting other hosts.

SkeletalElite commented:
Given that this lacks stealth, I think its fair. The moment nanites start being spread non willingly people the lynch mob will already start forming although I'm not sure how easy it is to get rid of your nanites once you have them because that will be necessary for the paranoid spacemen out there.

Nebulacrity commented:
> I'm not sure how easy it is to get rid of your nanites once you have them because that will be necessary for the paranoid spacemen out there.

EMPs, shocks, Ion Rifles and such help with that. Alternatively, raiding the lab and changing the nanite cloud's contents yourself is a valid option too.

oklews commented:
You could make it a stealth thing if you make it a tater item, would address a bit of the balance

Nebulacrity commented:
> You could make it a stealth thing if you make it a tater item, would address a bit of the balance

I think it's fine as is. It's unlocked via Military Nanites Programming research already, which requires a Syndicate item.

ExcessiveUseOfCobblestone commented on code/modules/research/nanites/nanite_programs/utility.dm:
```diff
@@ -258,6 +258,33 @@
+	if(!..())
+		return
+	var/list/mob/living/target_hosts = list()
+	for(var/mob/living/L in oview(1, host_mob))
```
code-wise this looks ok, only part i'm kinda scared about since it might get called quite a bit. I don't think it'll be too bad with just 1range but 🤷‍♂

Nebulacrity replied:
Keep in mind that the program has a cooldown. Shouldn't be that bad, right?

ExcessiveUseOfCobblestone replied:
depends on how many people are infected right? you send the signal and everyone with the bots does the thing?

Nebulacrity replied:
Yeah, but it's not like everyone's gonna be nearby many people all the time. Especially since there will definitely be people that'll notice the prick notification and will start yelling "ling!"

XDTM replied:
Check Adjacent, or this will be able to infect through window panes

Nebulacrity replied:
Done.

XDTM commented on code/modules/research/nanites/nanite_programs/utility.dm:
```diff
@@ -258,6 +258,33 @@
+/datum/nanite_program/triggered/nanite_sting
+	name = "Nanite Sting"
+	desc = "Stings a random non-host around the host with a barely-visible cluster of nanites, making them a new host. The target will feel it. \
+			If the cluster finds no valid targets, it returns to the original host."
```
"Projects a nearly invisible "spike" of nanites that attempts to infect a nearby non-host with a copy of the host's nanites cluster."

No need to give full details about the mechanics, they can be expanded in the wiki.

Nebulacrity replied:
Shortened the description.

Nebulacrity committed:
Added an adjacency check so the ability won't work on, for example, targets that are behind windows.
I have also shortened the description.
